csrc/device_lower/pass/index.cpp

```diff
@@ -611,10 +612,111 @@ void IndexLowering::handleBlockReduction(
   GpuLower::current()->propagateExprInfo(rop, back());
 }
 
+void IndexLowering::handleSerialGridReduction(
+    const ReductionOp* rop,
+    Val* out,
+    Val* in) {
+  const auto out_tv = out->as<kir::TensorIndex>()->view();
+  const auto out_domain = out_tv->domain();
+
+  // If we do a grid reduction we can't have a reduction axis that is not bound
+  // to a grid or block dim.
+  NVF_ERROR(
+      std::none_of(
+          out_domain->leaf().begin(),
+          out_domain->leaf().end(),
+          [](IterDomain* id) {
+            return !id->isThread() && id->isReduction() &&
+                !id->extent()->isOneInt();
+          }),
+      "Found a reduction stage that has both a non-parallelized ",
+      "reduction and a grid reduction. This is not supported, ",
+      "please use rfactor to do the serialized reduction first, ",
+      "then the grid reduction. ",
+      rop->toString());
+
+  NVF_ERROR(!rop->isAllreduce(), "Serial grid allReduce is not implemented");
+
+  // Allocate global work buffer TensorIndex.
```

Comment:
Why does the work buffer need to be scheduled? Can't it be simply just a buffer with the leaf domains of the reduction output tensor?

Reply:
I think you're right. It doesn't need to be scheduled explicitly, we just need to use the leaf domains of the output to determine how to index it (indexing as if the output were in global mem).

Reply:
Just pushed a change to simplify this by reusing `out_tv->domain()` instead of replaying to recreate it.

Reply:
This will share the same IterDomains with multiple TensorDomains, which I think we should avoid since we implicitly assume each tensor has unique IterDomains.

I think what we should do here would be something like:

```
std::vector<IterDomain*> work_buffer_ids(out_tv->nDims());
for (IterDomain* out_id : out_tv->leaf()) {
   work_buffer_ids.push_back(IterDomainBuilder(out_id).build());
}
auto work_buffer_domain = IrBuilder::create<TensorDomain>(work_buffer_ids);
```

Reply:
Ah OK. I see what you mean now. I was trying to preserve the transforms originally so that we could potentially re-use a a fusion output buffer for the work buffer in the future. However, that is probably low priority since it only applies when the output has the same type as the mma accumulator and we commonly have single precision accumulator and half-precision output tensors. For simplicity's sake I'll do as you suggest and just use the leaf domain of the mma output as the work buffer's root/leaf. I'll leave a comment to remind myself to revisit this if we implement global buffer re-use in the future.

Reply:
Sorry I wasn't clear. Yes, this should be enough for now.